### DISCUSSION TITLE
Kubernetes - Go - pulumi.String malformed

### DISCUSSION THREAD

**Discussion by Edregol** [pulumi/pulumi#16849] [Pulumi in Practice] [answered]
Pulumi Version: v3.127.0
Go Version: 1.22.5 windows/amd64

Hi everyone!
Hope you can help me out here. I'm using pulumi to deploy to a kubernetes cluster with go. More specifically, I'm using `github.com/pulumi/pulumi-kubernetes/sdk/v4/go/kubernetes/helm/v3` to deploy helm charts.

My issue is that when I provide a database connection string, like so: 
`"dsn": pulumi.String("postgresql://root@cockroach-cockroachdb-public:26257?sslmode=verify-full&sslrootcert=/cockroach-certs/ca.crt&sslcert=/cockroach-certs/client.root.crt&sslkey=/cockroach-certs/client.root.key")`

the service using it receives the string like this:

`postgresql://root@cockroach-cockroachdb-public:26257?sslcert=%2Fcockroach-certs%2Fclient.root.crt&sslkey=%2Fcockroach-certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=%2Fcockroach-certs%2Fca.crt`

You might see that sections of the string have been URL encoded which is not what I want here. Any Idea what I can do to fix this?

**Snippet:** 
```
_, err = helmv3.NewRelease(ctx, "kratos", &helmv3.ReleaseArgs{
  Name:      pulumi.String("kratos"),
  Chart:     pulumi.String(KRATOS_CHART_NAME),
  Namespace: authNamespaceName,
  RepositoryOpts: &helmv3.RepositoryOptsArgs{
    Repo: pulumi.String(KRATOS_CHART_URL),
  },
  Version: pulumi.String(KRATOS_CHART_VERSION),
  Values: pulumi.Map{
    "kratos": pulumi.Map{
      "config": pulumi.Map{
        "dsn": pulumi.String("postgresql://root@cockroach-cockroachdb-public:26257?sslmode=verify-full&sslrootcert=/cockroach-certs/ca.crt&sslcert=/cockroach-certs/client.root.crt&sslkey=/cockroach-certs/client.root.key"),
        "extraVolumes": pulumi.MapArray{
          pulumi.Map{
            "name": pulumi.String("cockroach-certs"),
            "secret": pulumi.Map{
              "secretName": pulumi.String("cockroach-cockroachdb-root-client"),
            }, 
          },
        },
        "extraVolumeMounts": pulumi.MapArray{
          pulumi.Map{
            "name": pulumi.String("cockroach-certs"),
            "mountPath": pulumi.String("/cockroach-certs"),
            "readOnly": pulumi.Bool(true),
          },
        },
      },
    },
})
```

Thanks for your help!

Cheers!

**Comment by Edregol** [comment]
Just realized that it might be the using service that alters the connection url since its attributes have been reordered.
Can I somehow check what the string looks like, when Pulumi creates the output for this?

**Reply by Frassle** [reply]
> Can I somehow check what the string looks like, when Pulumi creates the output for this?

Details view should show the full values for every resource iirc. Failing that, run with debug logs (--v10 --logtostderr) will definitely print the string value being sent to the provider at some point.

At any rate I suspect this to be an issue in the Kubernetes provider, or your Kubernetes service itself. I don't think the pulumi sdk and engine does any url encoding.

**Reply by Edregol** [reply]
Seems like pulumi and helmv3 are not doing anything to the string according to the logs:
`Unmarshaling property for RPC[Provider[kubernetes:helmrelease].Check(urn:pulumi:dev::infrastructure-deployment::kubernetes:helm.sh/v3:Release$kubernetes:core/v1:Namespace$kubernetes:helm.sh/v3:Release$kubernetes:core/v1:Namespace$kubernetes:helm.sh/v3:Release$kubernetes:cert-manager.io/v1:Issuer$kubernetes:helm.sh/v3:Release$kubernetes:helm.sh/v3:Release::kratos).news]: dsn={postgresql://root@cockroach-cockroachdb-public:26257?sslmode=verify-full&sslrootcert=/cockroach-certs/ca.crt&sslcert=/cockroach-certs/client.root.crt&sslkey=/cockroach-certs/client.root.key}`

I'll keep investigating. When I find a solution should I post it here or should I mark the question as resolved, since it is quite likely not related to pulumi?

**Reply by Frassle** [reply]
Why not post it here, might help somebody else :)

**Reply by Edregol** [reply, marked as answer]
Okay turns out I got distracted by the error message "unable to parse" in my service. The actual problem was a volume that I had not configured correctly. Which resulted in a "file not found" error when the service established a connection and tried to access the certificates paths.
In my defence though: I find it extremely confusing to post an error message that starts with "could not parse" and mentions somewhen at the end of the error "Um actually I could parse that string, but I couldn't find the file you are pointing to!"😆
